internal/controller/progressive/progressive.go

```diff
@@ -405,19 +406,30 @@ func AssessUpgradingPipelineType(
 
 	if analysisStatus != nil {
 		if upgradingObjectStatus.Phase == "Running" && healthyConditions && healthyReplicas && analysisStatus.Phase == argorolloutsv1.AnalysisPhaseSuccessful {
-			return apiv1.AssessmentResultSuccess, nil
+			return apiv1.AssessmentResultSuccess, "", nil
 		}
 	}
 
 	if upgradingObjectStatus.Phase == "Running" && healthyConditions && healthyReplicas {
-		return apiv1.AssessmentResultSuccess, nil
+		return apiv1.AssessmentResultSuccess, "", nil
 	}
 
 	if upgradingObjectStatus.Phase == "Failed" || !healthyConditions || !healthyReplicas {
-		return apiv1.AssessmentResultFailure, nil
+		failureReason := CalculateFailureReason(replicasFailureReason, upgradingObjectStatus.Phase, conditionFailureReason)
+		return apiv1.AssessmentResultFailure, failureReason, nil
 	}
 
-	return apiv1.AssessmentResultUnknown, nil
+	return apiv1.AssessmentResultUnknown, "", nil
+}
+
+func CalculateFailureReason(replicasFailureReason, phase, conditionFailureReason string) apiv1.AssessmentFailureReason {
+	if replicasFailureReason != "" {
+		return apiv1.AssessmentFailureReason(replicasFailureReason)
+	} else if phase == "Failed" {
+		return "child phase is in Failed state"
+	} else {
+		return apiv1.AssessmentFailureReason(conditionFailureReason)
+	}
```

Comment:
Let's reorder these, such that:
- phase == "Failed" is first
- condition Failure is second
- replicas is last